File: platforms/ios/Sources/Ratex/RaTeXEngine.swift
// RaTeXEngine.swift — Swift wrapper around the ratex_parse_and_layout C ABI.

import Foundation
import RaTeXFFI

// MARK: - Error type

public enum RaTeXError: Error, LocalizedError {
    case parseError(String)
    case nullResult

    public var errorDescription: String? {
        switch self {
        case .parseError(let msg): return "RaTeX parse error: \(msg)"
        case .nullResult:          return "RaTeX returned null with no error message"
        }
    }
}

// MARK: - Engine

/// Thread-safe entry point for RaTeX rendering.
///
/// ```swift
/// let displayList = try RaTeXEngine.shared.parse(#"\frac{-b \pm \sqrt{b^2-4ac}}{2a}"#)
/// ```
public final class RaTeXEngine {
    public static let shared = RaTeXEngine()
    private init() {}

    /// Parse a LaTeX string and return the corresponding `DisplayList`.
    ///
    /// This call is synchronous and CPU-bound; run it on a background queue for
    /// complex formulas.
    ///
    /// - Parameters:
    ///   - latex: A LaTeX math-mode string, e.g. `\frac{1}{2}`.
    ///   - displayMode: `true` (default) for display/block style (`$$...$$`);
    ///     `false` for inline/text style (`$...$`).
    /// - Returns: A `DisplayList` ready to be drawn.
    /// - Throws: `RaTeXError.parseError` on invalid LaTeX syntax.
    public func parse(_ latex: String, displayMode: Bool = true) throws -> DisplayList {
        var opts = RatexOptions(
            struct_size: MemoryLayout<RatexOptions>.size,
            display_mode: displayMode ? 1 : 0
        )
        let result = ratex_parse_and_layout(latex, &opts)
        guard result.error_code == 0, let ptr = result.data else {
            let msg: String
            if let errPtr = ratex_get_last_error() {
                msg = String(cString: errPtr)
            } else {
                msg = "unknown error"
            }
            throw RaTeXError.parseError(msg)
        }
        defer { ratex_free_display_list(ptr) }

        let json = String(cString: ptr)
        do {
            return try JSONDecoder().decode(DisplayList.self, from: Data(json.utf8))
        } catch {
            throw RaTeXError.parseError("JSON decode failed: \(error)")
        }
    }
}

File: platforms/jvm/src/main/kotlin/io/ratex/RaTeXEngine.kt
// RaTeXEngine.kt — JVM wrapper around libratex_ffi via JNA

package io.ratex

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

// MARK: - Error type

class RaTeXException(message: String) : Exception(message)

// MARK: - Engine

/**
 * Entry point for RaTeX rendering on the JVM.
 *
 * ```kotlin
 * val displayList = RaTeXEngine.parse("""\frac{-b \pm \sqrt{b^2-4ac}}{2a}""")
 * ```
 *
 * Note: [parse] is a suspend function; call it from a coroutine.
 * For one-shot calls from non-coroutine code, use [parseBlocking].
 */
object RaTeXEngine {

    private val native: RaTeXNative = RaTeXNative.INSTANCE

    /**
     * Parse [latex] and return a [DisplayList] decoded from the JSON result.
     * Runs on [Dispatchers.Default].
     *
     * @throws RaTeXException on parse or decode error.
     */
    suspend fun parse(latex: String): DisplayList = withContext(Dispatchers.Default) {
        parseBlocking(latex)
    }

    /**
     * Blocking variant of [parse]. Safe to call on any thread.
     *
     * @throws RaTeXException on parse or decode error.
     */
    fun parseBlocking(latex: String): DisplayList {
        val ptr = native.ratex_parse_and_layout(latex)
            ?: throw RaTeXException(native.ratex_get_last_error() ?: "unknown error")
        val json: String
        try {
            json = ptr.getString(0, "UTF-8")
        } finally {
            native.ratex_free_display_list(ptr)
        }
        return try {
            ratexJson.decodeFromString(DisplayList.serializer(), json)
        } catch (e: Exception) {
            throw RaTeXException("JSON decode failed: ${e.message}")
        }
    }
}
